crucible-go/src/Lang/Crucible/Go/Types.hs

```diff
@@ -0,0 +1,326 @@
+
+type instance ExprExtension Go = EmptyExprExtension
+type instance StmtExtension Go = EmptyStmtExtension
+
+instance IsSyntaxExtension Go
+
+goIntrinsicTypes :: IntrinsicTypes sym
+goIntrinsicTypes = emptyIntrinsicTypes
+
+type Verbosity = Int
+
+-- | A function-local register.
+data GoReg s tp where
+  GoReg :: TypeRepr tp -> Gen.Reg s (ReferenceType tp) -> GoReg s tp
+  deriving Show
+
+type SomeGoReg s = Some (GoReg s)
+
+data GoGlobal where
+  GoGlobal :: Gen.GlobalVar tp -- ^ the global variable identifier
+           -> (forall s. Gen.Expr Go s tp) -- ^ initial (zero) value
+           -> GoGlobal
+deriving instance Show GoGlobal
+
+-- | As assignable location. Either a statically known location
+-- (reference, global, array offset) or a pointer expression.
+data GoLoc s tp where
+  GoLocRef :: Gen.Expr Go s (ReferenceType tp)
+           -> GoLoc s tp
+  GoLocGlobal :: Gen.GlobalVar tp
+              -> GoLoc s tp
+  GoLocArray :: Gen.Expr Go s (ReferenceType (VectorType tp))
+             -> Gen.Expr Go s NatType
+             -> GoLoc s tp
+  GoLocPointer :: Gen.Expr Go s (PointerType tp)
+               -> GoLoc s tp
+  deriving Show
+
+-- | Existentially packaged positive natural number.
+data PosNat = forall w. PosNat (NatRepr w) (LeqProof 1 w)
+
+intToPosNat :: Int -> Maybe PosNat
+intToPosNat i = do
+  Some w <- someNat (fromIntegral i)
+  LeqProof <- isPosNat w
+  return $ PosNat w LeqProof
+
+data AnyGoExpr where
+  AnyGoExpr :: (forall s. SomeGoExpr s) -> AnyGoExpr
+deriving instance Show AnyGoExpr
+
+-- | A package environment. Note that Go is a "lisp-1" language in the
+-- sense that variables and functions share the same namespace, but we
+-- keep them separately 1) for convenience and 2) globals can't be
+-- embedded in expressions.
+data Namespace = Namespace { ns_globals :: HashMap Text GoGlobal
+                           , ns_funcs :: HashMap Text SomeHandle
+                           }
+  deriving Show
+
+insert_global :: Text -> GoGlobal -> Namespace -> Namespace
+insert_global name glob ns =
+  ns { ns_globals = HM.insert name glob (ns_globals ns) }
+
+insert_function :: Text -> SomeHandle -> Namespace -> Namespace
+insert_function name h ns =
+  ns { ns_funcs = HM.insert name h (ns_funcs ns) }
+
+instance Default Namespace where
+  def = Namespace { ns_globals = HM.empty
+                  , ns_funcs = HM.empty }
+
+namespace_union :: Namespace -> Namespace -> Namespace
+namespace_union (Namespace g1 f1) (Namespace g2 f2) =
+  Namespace { ns_globals = HM.union g1 g2
+            , ns_funcs = HM.union f1 f2 }
+
+-- | Translator state.
+data TransState =
+  TransState
+  { machineWordWidth :: PosNat -- ^ Bit width of unqualified "int" and
+                               -- "uint" types
+  , retRepr :: Some TypeRepr -- ^ Return type of current function
+  , namespaces :: HashMap Text Namespace
+  , pkgName :: Text -- ^ Name of package curently being translated
+  , sng :: Maybe (Some (NonceGenerator IO))
```

Comment:
Just a note to myself - I'm surprised to see this `NonceGenerator` have its state thread type parameter quantified away.  I wonder how it is used such that this is okay.  It seems like any use wouldn't really be possible unless one didn't actually need a nonce.

Reply:
Ah, I see - it is passed to the Crucible Generator runner, which uses an arbitrary nonce generator and quantifies away the `s`.  We could just get rid of this and pass the `globalIONonceGenerator` in the Translation module

Reply:
Fixed.